synapse/handlers/oidc.py

```diff
@@ -1043,6 +1213,180 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
         # to be strings.
         return str(remote_user_id)
 
+    async def handle_backchannel_logout(
+        self, request: SynapseRequest, logout_token: str
+    ) -> None:
+        """Handle an incoming request to /_synapse/client/oidc/backchannel_logout
+
+        The OIDC Provider posts a logout token to this endpoint when a user
+        session ends. That token is a JWT signed with the same keys as
+        ID tokens. The OpenID Connect Back-Channel Logout draft explains how to
+        validate the JWT and figure out what session to end.
+
+        Args:
+            request: The request to respond to
+            logout_token: The logout token (a JWT) extracted from the request body
+        """
+        # Back-Channel Logout can be disabled in the config, hence this check.
+        # This is not that important for now since Synapse is registered
+        # manually to the OP, so not specifying the backchannel-logout URI is
+        # as effective than disabling it here. It might make more sense if we
+        # support dynamic registration in Synapse at some point.
+        if not self._config.backchannel_logout_enabled:
+            logger.warning(
+                f"Received an OIDC Back-Channel Logout request from issuer {self.issuer!r} but it is disabled in config"
+            )
+            raise SynapseError(
+                400, "OpenID Connect Back-Channel Logout is disabled for this provider"
+            )
```

Comment:
This will get spit back out as a JSON error to the IdP -- is that an appropriate response?

Reply:
> If the logout succeeded, the RP MUST respond with HTTP 200 OK. However, note that some Web frameworks will substitute an HTTP 204 No Content response for an HTTP 200 OK when the HTTP body is empty. Therefore, OPs should be prepared to also process an HTTP 204 No Content response as a successful response.
>
> If the logout request was invalid or the logout failed, the RP **MUST respond with HTTP 400 Bad Request**. The response **MAY include an HTTP body consisting of a JSON object with error and error_description** parameters conveying the nature of the error that occured, which can assist with debugging. These error response parameters are used as specified in Section 5.2 of OAuth 2.0 [RFC6749]. Like in OAuth 2.0, the parameters are included in the entity-body of the HTTP response using the application/json media type. Also like in OAuth 2.0, the error parameter is REQUIRED when a response body is present and the error_description parameter is OPTIONAL. An error value of invalid_request MAY be used to indicate that there was a problem with the syntax of the logout request. Note that the information conveyed in the response body is intended to help debug deployments; it is not intended that implementations use different error values to trigger different runtime behaviors

So the 400 code is a MUST, the JSON with error + error_description is a SHOULD, soooo this is probably fine?

Reply:
That sounds OK. Can we add a todo saying something about filling in error / error_description please.